src/kernel/arch/aarch64/arch.zig

```diff
@@ -35,50 +36,38 @@ pub fn initTTY(boot_payload: BootPayload) TTY {
 
 pub fn initSerial(board: BootPayload) Serial {
     mmio_addr = board.mmioAddress();
-    // Disable uart
-    mmio.write(mmio_addr, mmio.Register.UART_CONTROL, 0);
-    // Disable pull up/down for all GPIO pins
-    mmio.write(mmio_addr, mmio.Register.GPIO_PULL, 0);
-    // Disable pull up/down for pins 14 and 15
-    mmio.write(mmio_addr, mmio.Register.GPIO_PULL_CLK, (1 << 14) | (1 << 15));
-    // Apply pull up/down change for pins 14 and 15
-    mmio.write(mmio_addr, mmio.Register.GPIO_PULL_CLK, 0);
-    // Clear pending uart interrupts
-    mmio.write(mmio_addr, mmio.Register.UART_INT_CONTROL, 0x7FF);
-
-    // For models after rpi 2 (those supporting aarch64) the uart clock is dependent on the system clock
-    // so set it to a known constant rather than calculating it
-    const cmd: []const volatile u32 align(16) = &[_]u32{ 36, 0, 0x38002, 12, 8, 2, SYSTEM_CLOCK, 0, 0 };
-    const cmd_addr = (@intCast(u32, @ptrToInt(&cmd)) & ~@as(u32, 0xF)) | 8;
-    // Wait until the mailbox isn't busy
-    while ((mmio.read(mmio_addr, mmio.Register.MBOX_STATUS) & 0x80000000) != 0) {}
-    mmio.write(mmio_addr, mmio.Register.MBOX_WRITE, cmd_addr);
-    // Wait for the correct response
-    while ((mmio.read(mmio_addr, mmio.Register.MBOX_STATUS) & 0x40000000) != 0 or mmio.read(mmio_addr, mmio.Register.MBOX_READ) != cmd_addr) {}
-
-    // Setup baudrate
-    const divisor: f32 = @intToFloat(f32, SYSTEM_CLOCK) / @intToFloat(f32, 16 * UART_BAUD_RATE);
-    const divisor_int: u32 = @floatToInt(u32, divisor);
-    const divisor_frac: u32 = @floatToInt(u32, (divisor - @intToFloat(f32, divisor_int)) * 64.0 + 0.5);
-    mmio.write(mmio_addr, mmio.Register.UART_BAUD_INT, divisor_int);
-    mmio.write(mmio_addr, mmio.Register.UART_BAUD_FRAC, divisor_frac);
-
-    // Enable FIFO, 8 bit words, 1 stop bit and no parity bit
-    mmio.write(mmio_addr, mmio.Register.UART_LINE_CONTROL, 1 << 4 | 1 << 5 | 1 << 6);
-    // Mask all interrupts
-    mmio.write(mmio_addr, mmio.Register.UART_INT_MASK, 1 << 1 | 1 << 4 | 1 << 5 | 1 << 6 | 1 << 7 | 1 << 8 | 1 << 9 | 1 << 10);
-    // Enable UART0 receive and transmit
-    mmio.write(mmio_addr, mmio.Register.UART_CONTROL, 1 << 0 | 1 << 8 | 1 << 9);
-
+    is_qemu = interrupts.mrs("cntfrq_el0") != 0;
+    if (!is_qemu) {
+        pinSetFunction(14, .AlternateFunction5);
+        pinSetPull(14, .None);
+        pinSetFunction(15, .AlternateFunction5);
+        pinSetPull(15, .None);
+        mmio.write(mmio_addr, .AUX_ENABLES, 1);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_LCR_REG, 3);
+        mmio.write(mmio_addr, .AUX_MU_MCR_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IIR_REG, 0xc6);
+        mmio.write(mmio_addr, .AUX_MU_BAUD_REG, 270);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 3);
+    }
```

Comment:
Why did all of this need to be changed? All the comments, system clock configuration and such has been removed.

Reply:
Uart0 is the full uart and uart1 is the mini (aux) uart. Uart0 is used for the bluetooth chip on 3b and above, leaving uart1 for anything else like the serial console.

However, qemu does not implement uart1, it only implements uart0, and of course it does not implement the bluetooth chip.

Therefore uart1 is used for the console except when under qemu. Under qemu uart0 is used for the console. Also qemu uart0 does not require initialization so uart0 constants were removed. Uart0 constants and initialization would be added back, probably differently, for bluetooth when we get to it.

By the way, the uart0 interface to the bluetooth chip is quite elegant and easy to program.

Reply:
I see. So does uart1 not depend on the system clock? This code could do with some comments to explain what is happening.

Reply:
I have added comments regarding the use of uart0/uart1 in relation to qemu and also to the system clock. enable_uart=1 is now included in config.txt to ensure the baud rate is stable.